File: packages/surface/src/surfaceFrom.ts
import { createEntity } from '@flighthq/entity';
import { createImageResourceFromCanvas } from '@flighthq/image';
import type { ImageResource, Surface } from '@flighthq/types';

export function createImageResourceFromSurface(surface: Readonly<Surface>): ImageResource {
  const canvas = document.createElement('canvas');
  canvas.width = surface.width;
  canvas.height = surface.height;
  const domImageData = new globalThis.ImageData(surface.width, surface.height);
  domImageData.data.set(surface.data);
  canvas.getContext('2d')!.putImageData(domImageData, 0, 0);
  return createImageResourceFromCanvas(canvas);
}

export function createSurfaceFromCanvas(
  canvas: HTMLCanvasElement,
  x: number = 0,
  y: number = 0,
  width?: number,
  height?: number,
): Surface {
  const w = width ?? canvas.width;
  const h = height ?? canvas.height;
  const ctx = canvas.getContext('2d')!;
  const raw = ctx.getImageData(x, y, w, h);
  return createEntity({
    alphaType: 'straight',
    colorSpace: raw.colorSpace as 'srgb' | 'display-p3',
    compressed: null,
    data: raw.data,
    format: 'rgba8unorm',
    height: raw.height,
    source: null,
    version: 0,
    width: raw.width,
  });
}

export function createSurfaceFromImageResource(resource: Readonly<ImageResource>): Surface {
  const canvas = document.createElement('canvas');
  canvas.width = resource.width;
  canvas.height = resource.height;
  if (resource.source === null) {
    // Data-only resource: copy its existing pixels when present, otherwise allocate a transparent buffer.
    return createEntity({
      alphaType: resource.alphaType,
      colorSpace: 'srgb' as const,
      compressed: null,
      data:
        resource.data !== null
          ? new Uint8ClampedArray(resource.data)
          : new Uint8ClampedArray(resource.width * resource.height * 4),
      format: resource.format,
      height: resource.height,
      source: null,
      version: 0,
      width: resource.width,
    });
  }
  const ctx = canvas.getContext('2d')!;
  ctx.drawImage(resource.source, 0, 0);
  const raw = ctx.getImageData(0, 0, resource.width, resource.height);
  return createEntity({
    alphaType: 'straight',
    colorSpace: raw.colorSpace as 'srgb' | 'display-p3',
    compressed: null,
    data: raw.data,
    format: 'rgba8unorm',
    height: resource.height,
    source: null,
    version: 0,
    width: resource.width,
  });
}

/**
 * Reads a surface out of any `CanvasImageSource` by drawing it into a scratch 2D canvas. Unlike
 * createSurfaceFromCanvas — which calls getContext('2d') and so only works on a 2D-rendered canvas —
 * this captures a Gl or Wgpu canvas too, giving one readback path for every render backend.
 * `width`/`height` are the device-pixel dimensions to capture (pass the render state's canvas size).
 * For a Gl/Wgpu source, draw before the browser composites the frame away (in tests, enable the
 * context's preserveDrawingBuffer, or read immediately after rendering).
 */
export function createSurfaceFromImageSource(source: CanvasImageSource, width: number, height: number): Surface {
  const canvas = document.createElement('canvas');
  canvas.width = width;
  canvas.height = height;
  const ctx = canvas.getContext('2d')!;
  ctx.drawImage(source, 0, 0);
  const raw = ctx.getImageData(0, 0, width, height);
  return createEntity({
    alphaType: 'straight',
    colorSpace: raw.colorSpace as 'srgb' | 'display-p3',
    compressed: null,
    data: raw.data,
    format: 'rgba8unorm',
    height,
    source: null,
    version: 0,
    width,
  });
}

File: crates/flighthq-surface/src/surface_from.rs
//! Surface construction from other resource types.

use flighthq_types::{AlphaType, ColorSpace, ImageResource, PixelFormat, Surface};

/// Converts a `Surface` into an `ImageResource`. On native the pixel buffer
/// is copied; there is no canvas element.
pub fn create_image_resource_from_surface(surface: &Surface) -> ImageResource {
    ImageResource {
        alpha_type: surface.alpha_type,
        data: Some(surface.data.clone()),
        format: surface.format,
        height: surface.height,
        version: 0,
        width: surface.width,
    }
}

/// Constructs a `Surface` from an existing `ImageResource`. If the resource
/// has CPU pixel data (`data` is `Some`), it is copied; otherwise returns a
/// transparent surface of the same dimensions.
pub fn create_surface_from_image_resource(resource: &ImageResource) -> Surface {
    let data = match &resource.data {
        Some(d) => d.clone(),
        None => vec![0u8; (resource.width as usize) * (resource.height as usize) * 4],
    };
    Surface {
        alpha_type: resource.alpha_type,
        color_space: ColorSpace::Srgb,
        data,
        format: resource.format,
        height: resource.height,
        version: 0,
        width: resource.width,
    }
}

/// Constructs a `width`x`height` `Surface` by capturing an image byte source.
///
/// On the web, `createSurfaceFromImageSource` draws any `CanvasImageSource`
/// (including a WebGL/WebGPU canvas) into a scratch 2D canvas at `(0, 0)` and
/// reads back a `width`x`height` region — one readback path for every render
/// backend. There is no canvas on native, so the byte-source analogue copies
/// the `source` resource's pixels into the top-left of a fresh transparent
/// `width`x`height` surface, clipping anything past the requested size. This
/// mirrors the web path's `drawImage(source, 0, 0)` followed by a full-frame
/// `getImageData(0, 0, width, height)`.
///
/// The `source.data` must be RGBA8 (`PixelFormat::Rgba8Unorm`); an element-only
/// resource (`data` is `None`) yields a fully transparent surface.
pub fn create_surface_from_image_source(
    source: &ImageResource,
    width: u32,
    height: u32,
) -> Surface {
    let mut out = vec![0u8; (width as usize) * (height as usize) * 4];
    if let Some(src) = source.data.as_ref() {
        let copy_w = source.width.min(width);
        let copy_h = source.height.min(height);
        for y in 0..copy_h {
            let src_row = (y * source.width) as usize * 4;
            let dst_row = (y * width) as usize * 4;
            let row_bytes = copy_w as usize * 4;
            out[dst_row..dst_row + row_bytes].copy_from_slice(&src[src_row..src_row + row_bytes]);
        }
    }
    Surface {
        alpha_type: AlphaType::Straight,
        color_space: ColorSpace::Srgb,
        data: out,
        format: PixelFormat::Rgba8Unorm,
        height,
        version: 0,
        width,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::surface::create_surface;

    #[test]
    fn create_image_resource_from_surface_copies_pixels() {
        let surface = create_surface(2, 2, 0x11223344);
        let resource = create_image_resource_from_surface(&surface);
        assert_eq!(resource.width, 2);
        assert_eq!(resource.height, 2);
        assert_eq!(resource.data.as_ref().unwrap(), &surface.data);
        assert_eq!(resource.version, 0);
    }

    #[test]
    fn create_surface_from_image_resource_data_present() {
        let surface = create_surface(2, 1, 0x55667788);
        let resource = create_image_resource_from_surface(&surface);
        let back = create_surface_from_image_resource(&resource);
        assert_eq!(back.width, 2);
        assert_eq!(back.height, 1);
        assert_eq!(back.data, surface.data);
    }

    #[test]
    fn create_surface_from_image_resource_no_data() {
        let resource = ImageResource {
            data: None,
            height: 3,
            width: 4,
            ..ImageResource::default()
        };
        let surface = create_surface_from_image_resource(&resource);
        assert_eq!(surface.data.len(), 4 * 3 * 4);
        assert!(surface.data.iter().all(|&b| b == 0));
    }

    #[test]
    fn create_surface_from_image_source_captures_at_the_given_device_size() {
        let source = create_image_resource_from_surface(&create_surface(8, 4, 0x112233ff));
        let surface = create_surface_from_image_source(&source, 8, 4);
        assert_eq!(surface.width, 8);
        assert_eq!(surface.height, 4);
        assert_eq!(surface.data.len(), 8 * 4 * 4);
        // The captured pixels match the source resource.
        assert_eq!(surface.data, *source.data.as_ref().unwrap());
    }

    #[test]
    fn create_surface_from_image_source_clips_a_larger_source_to_the_top_left() {
        let source = create_image_resource_from_surface(&create_surface(4, 4, 0xaabbccff));
        let surface = create_surface_from_image_source(&source, 2, 2);
        assert_eq!(surface.width, 2);
        assert_eq!(surface.height, 2);
        // Every captured pixel is the opaque source color.
        for px in surface.data.chunks_exact(4) {
            assert_eq!(px, [0xaa, 0xbb, 0xcc, 0xff]);
        }
    }

    #[test]
    fn create_surface_from_image_source_leaves_uncovered_area_transparent() {
        let source = create_image_resource_from_surface(&create_surface(2, 2, 0xffffffff));
        let surface = create_surface_from_image_source(&source, 4, 4);
        // Top-left 2x2 is opaque white; the rest stays transparent black.
        assert_eq!(&surface.data[0..4], &[0xff, 0xff, 0xff, 0xff]);
        // Last pixel (bottom-right) is uncovered.
        let last = surface.data.len() - 4;
        assert_eq!(&surface.data[last..], &[0, 0, 0, 0]);
    }

    #[test]
    fn create_surface_from_image_source_element_only_is_transparent() {
        let source = ImageResource {
            data: None,
            width: 4,
            height: 4,
            ..ImageResource::default()
        };
        let surface = create_surface_from_image_source(&source, 3, 2);
        assert_eq!(surface.data.len(), 3 * 2 * 4);
        assert!(surface.data.iter().all(|&b| b == 0));
    }
}
